app/controllers/VersionSpecificationController.java

```diff
@@ -0,0 +1,329 @@
+
+/**
+ * Metrics portal version specifications controller. Exposes APIs to add, remove and query version specifications and
+ * associated package-versions.
+ *
+ * @author Matthew Hayter (mhayter at groupon dot com)
+ */
+@Singleton
+public class VersionSpecificationController extends Controller {
+
+    /**
+     * Public constructor.
+     *
+     * @param configuration Instance of Play's <code>Configuration</code>.
+     * @param versionSpecificationRepository Instance of <code>AlertRepository</code>.
+     */
+    @Inject
+    public VersionSpecificationController(
+            final Configuration configuration,
+            final VersionSpecificationRepository versionSpecificationRepository) {
+        this(configuration.getInt("version_specifications.limit", DEFAULT_MAX_LIMIT), versionSpecificationRepository);
+    }
+
+    /**
+     * The primary endpoint for Evergreen. Given a set of host attributes, this endpoint returns the
+     * appropriate version set as determined by the set of <code>VersionSpecifications</code> and their
+     * associated <code>VersionSets</code>. See the README.md for more information.
+     *
+     * @return <code>Result</code> VersionSet
+     */
+    public Result lookupHostVersionSet() {
+
+        // If the queryString contains multiple values for a single parameter, select the last parameter.
+        final Map<String, String> hostAttributes = selectLastQueryParams(request().queryString());
+
+        final Optional<models.internal.VersionSet> versionSet =
+                _versionSpecificationRepository.lookupVersionSetByHostAttributes(hostAttributes);
+
+        if (!versionSet.isPresent()) {
+            // This is an error in the Evergreen dataset: there should always be a default/fallback version set specified.
+            LOGGER.error()
+                    .setMessage("No VersionSet found for host attributes")
+                    .addData("hostAttributes", hostAttributes)
+                    .log();
+            return internalServerError();
+        }
+        return ok(Json.toJson(internalToViewModel(versionSet.get())));
+    }
+
+    /**
+     * Get the <code>VersionSet</code> associated with the given UUID.
+     *
+     * @param uuid The UUID.
+     * @return <code>Result</code> version set.
+     */
+    public Result getVersionSet(final UUID uuid) {
+        final Optional<models.internal.VersionSet> maybeVersionSet = _versionSpecificationRepository.getVersionSet(uuid);
+        if (!maybeVersionSet.isPresent()) {
+            return notFound();
+        }
+        return ok(Json.toJson(internalToViewModel(maybeVersionSet.get())));
+    }
+
+    /**
+     * Update or create a VersionSet with the given UUID.
+     *
+     * @param uuid The UUID.
+     * @return <code>Result</code> status code indicating success or failure.
+     */
+    public Result updateCreateVersionSet(final UUID uuid) {
+        final VersionSet view = request().body().as(VersionSet.class);
```

Comment:
We generally avoid the use of the Play built-in JSON conversion routines and rely on our own ObjectMapper instance. Take a look at Deepika's expressions and alerts controllers.